### ISSUE TITLE
Unallowed modules not unloaded in sanity check

### ISSUE THREAD

**Issue by Flamefire** [easybuilders/easybuild-framework#3456]
For the missing ncurses dependency fixed in https://github.com/easybuilders/easybuild-easyconfigs/pull/11415 I've seen that the sanity check checking `clang++ --help` passes if I have ncurses loaded when running EB, e.g. through loading `git`.

So it seems that for the sanity check step the disallowed modules are not unloaded which could be the reason why this went unnoticed for so long (besides the missing sanity check commands)

**Comment by boegel** [comment]
@Flamefire With unallowed modules you mean modules that are loaded and not listed in `--allow-loaded-modules` but which were loaded manually before running `eb`?

That's (part of) the reason why we produce a warning for these modules in the first place, because they may affect the installation somehow.

I don't think there's anything we can do about this... Blindly unloading those modules when running the sanity check may cause the sanity check to fail, while it's possible that the modules were loaded exactly to try and get the sanity check to pass (which is a shoot-yourself-in-the-foot scenario, I guess).

Why did you have `ncurses`/`git` loaded through a module?

**Comment by Flamefire** [comment]
Yes. I have `detect-loaded-modules=unload` set, so those are usually unloaded but it seems they get restored in the sanity check step. And this is what I consider a bug.
IMO "unload" that would have been the better default in the first place, but I guess that's to late.

Well this is usually when developing ECs: The system git is very old and therefore slow and missing e.g. auto-complete. Hence I load a git module, work on the EC tree and build/test there relying on EB unloading those extra modules during the build

**Comment by boegel** [comment]
Thanks for clarifying, it's indeed a bug that those modules are restored during the sanity check...

**Comment by boegel** [comment]
I took a closer look at this, and based on the logic in the code I don't really see how the loaded modules can get restored during the sanity check step...

We take a snapshot of the environment right before we start each installation (see the `init_env = ...` in `build_and_install_software` in `main.py`), and that's done *after* we deal with already loaded modules (which is done by `modtool.check_loaded_modules()`, which unloads those modules if `unload` mode is used).

Perhaps the problem is that you're loading that `git` (+ `ncurses`) module in your `.bashrc` startup script, so in fact it's picked up every time we run a shell command?

**Comment by Flamefire** [comment]
No I did load it manually. However I wasn't able to reproduce this in a simple example, so no idea what went wrong there

**Comment by boegel** [comment]
@Flamefire Sould we keep this open, since you're not able to reproduce the issue?

**Comment by Flamefire** [comment]
I think we can close this and reopen when it reappears although its gonna be hard as it was only triggered by a faulty EC :/

Might be related to https://github.com/TACC/Lmod/issues/508 which I just filed where I had issues with unloading in LMod directly
